File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/util/scopedservices/ServiceProvider.java
package com.zhuinden.simplestackdemoexamplefragments.util.scopedservices;

import android.support.annotation.NonNull;

import com.zhuinden.simplestack.ScopedServices;

public class ServiceProvider implements ScopedServices {
    @Override
    public void bindServices(@NonNull ServiceBinder serviceBinder) {
        if(serviceBinder.getKey() instanceof HasServices) {
            ((HasServices) serviceBinder.getKey()).bindServices(serviceBinder);
        }
    }
}

File: tutorials/tutorial-sample/src/main/java/com/zhuinden/simplestacktutorials/steps/step_8/core/viewmodels/ServiceProvider.kt
package com.zhuinden.simplestacktutorials.steps.step_8.core.viewmodels

import com.zhuinden.simplestack.ScopedServices
import com.zhuinden.simplestack.ServiceBinder
import com.zhuinden.simplestacktutorials.steps.step_8.core.navigation.FragmentKey

class ServiceProvider : ScopedServices {
    override fun bindServices(serviceBinder: ServiceBinder) {
        val key = serviceBinder.getKey<FragmentKey>()

        val scope = serviceBinder.scopeTag

        if (key is HasServices && key.scopeTag == scope) {
            key.bindServices(serviceBinder) // screen-bound shared services
        }
    }
}